### PR TITLE
Zscaler plugin - urls and domains ad input parameter  in actions Blacklist URL and Lookup URL

### PR DESCRIPTION
**Pull request by r7-kszczepanskagorna**
## Proposed Changes
Zscaler plugin - urls and domains ad input parameter in actions Blacklist URL and Lookup URL

### Description

Describe the proposed changes:

  -

### Testing

Any testing information goes here.

## PR Requirements

Developers, verify you have completed the following items by checking them off:

### Style

Review the [style guide](https://komand.github.io/python/style.html)

- [ ] For dependencies, pin [OS package](https://komand.github.io/python/style.html#dockerfile) and [Python package](https://komand.github.io/python/style.html#requirements-txt) versions
- [ ] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [ ] For size, use the [slim SDK images](https://komand.github.io/python/sdk.html#sdk-versions) when possible: ``komand/python-3-37-slim-plugin`` and ``komand/python-3-37-plugin``
- [ ] For error handling, use of [PluginException](https://komand.github.io/python/error_handling.html#plugin-exceptions) and [ConnectionTestException](https://komand.github.io/python/error_handling.html#connection-exceptions)
- [ ] For logging, use [self.logger](https://komand.github.io/python/sdk.html#logging)
- [ ] For docs, use [changelog style](https://komand.github.io/python/style.html#changelog)
- [ ] For docs, validate markdown with ``make validate`` which calls ``mdl`` to lint ``help.md``

### Functional Checklist
- [ ] Work fully completed
- [ ] Functional
  - [ ] Any new actions/triggers include JSON [test files](https://komand.github.io/python/style.html#tests) in the `tests/` directory created with `icon-plugin run -c sample $action > tests/$action.json`
  - [ ] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -T all --debug --jq` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -R all --debug --jq` (use PR format at end)

### Run

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/Zscaler:1.1.2. Step name: blacklist_url\n",
    "meta": {},
    "output": {
      "success": true
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/zscaler:1.1.2 --debug run < tests/blacklist_url_add.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/Zscaler:1.1.2. Step name: blacklist_url\n",
    "meta": {},
    "output": {
      "success": true
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/zscaler:1.1.2 --debug run < tests/blacklist_url_remove.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "Connect: Connecting...\nrapid7/Zscaler:1.1.2. Step name: lookup_url\n",
    "meta": {},
    "output": {
      "url_categorization": [
        {
          "url": "google.com",
          "urlClassifications": [
            "WEB_SEARCH"
          ],
          "urlClassificationsWithSecurityAlert": []
        },
        {
          "url": "rapid7.com",
          "urlClassifications": [
            "CORPORATE_MARKETING",
            "PROFESSIONAL_SERVICES"
          ],
          "urlClassificationsWithSecurityAlert": []
        },
        {
          "url": "wp.pl",
          "urlClassifications": [
            "NEWS_AND_MEDIA"
          ],
          "urlClassificationsWithSecurityAlert": []
        }
      ]
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/zscaler:1.1.2 --debug run < tests/lookup_url.json
</summary>
</details>

<details>

```
[*] Validating plugin with all validators at .

[*] Running Integration Validators...
[*] Executing validator HelpValidator
[*] Executing validator ChangelogValidator
[*] Executing validator RequiredKeysValidator
[*] Executing validator UseCaseValidator
[*] Executing validator SpecPropertiesValidator
[*] Executing validator SpecVersionValidator
[*] Executing validator FilesValidator
[*] Executing validator TagValidator
[*] Executing validator DescriptionValidator
[*] Executing validator TitleValidator
[*] Executing validator VendorValidator
[*] Executing validator DefaultValueValidator
[*] Executing validator IconValidator
[*] Executing validator RequiredValidator
[*] Executing validator VersionValidator
[*] Executing validator DockerfileParentValidator
[*] Executing validator LoggingValidator
[*] Executing validator ProfanityValidator
[*] Executing validator AcronymValidator
[*] Executing validator JSONValidator
[*] Executing validator OutputValidator
[*] Executing validator RegenerationValidator
[*] Executing validator HelpInputOutputValidator
[*] Executing validator SupportValidator
[*] Executing validator RuntimeValidator
[*] Executing validator ExceptionValidator
[*] Executing validator CredentialsValidator
[*] Executing validator PasswordValidator
[*] Executing validator PrintValidator
[*] Executing validator ConfidentialValidator
[*] Executing validator DockerValidator
[*] Executing validator URLValidator
[*] Plugin successfully validated!

----
[*] Total time elapsed: 7752.009ms

```

<summary>
icon-validate --all .
</summary>
</details>

<details>

```
[*] Use ``make menu`` for available targets
[*] Including available Makefiles: ../tools/Makefiles/Helpers.mk ../tools/Makefiles/Colors.mk
--
[*] Running validators
[*] Validating plugin at .

[*] Running Integration Validators...
[*] Executing validator HelpValidator
[*] Executing validator ChangelogValidator
[*] Executing validator RequiredKeysValidator
[*] Executing validator UseCaseValidator
[*] Executing validator SpecPropertiesValidator
[*] Executing validator SpecVersionValidator
[*] Executing validator FilesValidator
[*] Executing validator TagValidator
[*] Executing validator DescriptionValidator
[*] Executing validator TitleValidator
[*] Executing validator VendorValidator
[*] Executing validator DefaultValueValidator
[*] Executing validator IconValidator
[*] Executing validator RequiredValidator
[*] Executing validator VersionValidator
[*] Executing validator DockerfileParentValidator
[*] Executing validator LoggingValidator
[*] Executing validator ProfanityValidator
[*] Executing validator AcronymValidator
[*] Executing validator JSONValidator
[*] Executing validator OutputValidator
[*] Executing validator RegenerationValidator
[*] Executing validator HelpInputOutputValidator
[*] Executing validator SupportValidator
[*] Executing validator RuntimeValidator
[*] Plugin successfully validated!

----
[*] Total time elapsed: 1076.375ms

[*] Validating spec with js-yaml
[SUCCESS] Passes js-yaml spec check


[*] Validating markdown...
[SUCCESS] Passes markdown linting

[*] Validating python files for style...
[SUCCESS] Passes flake8 linting

[*] Validating python files for security vulnerabilities...
[main]  INFO    profile include tests: None
[main]  INFO    profile exclude tests: None
[main]  INFO    cli include tests: None
[main]  INFO    cli exclude tests: None
[main]  INFO    running on Python 3.7.3
Run started:2020-09-29 20:51:50.564001

Test results:
        No issues identified.

Code scanned:
        Total lines of code: 429
        Total lines skipped (#nosec): 0

Run metrics:
        Total issues (by severity):
                Undefined: 0.0
                Low: 0.0
                Medium: 0.0
                High: 0.0
        Total issues (by confidence):
                Undefined: 0.0
                Low: 0.0
                Medium: 0.0
                High: 0.0
Files skipped (0):
[SUCCESS] Passes bandit security checks

[*] Checking for typos with Misspell...
[SUCCESS] Passes Misspell check

```

<summary>
make validate
</summary>
</details>

### CODE REVIEW

File: zscaler/help.md [hunk 1]
```diff
@@ -161,6 +161,7 @@ _This plugin does not contain any troubleshooting information._
 
 # Version History
 
+* 1.1.2 - Change input `URLs` in actions Blacklist URL and Lookup URL to accepts URls and domains
```
Comment (jschipp-r7) [inline zscaler/help.md:164]:
```suggestion
* 1.1.2 - Support both domains and URLs in the Blacklist URL and Lookup URL actions
```

File: zscaler/icon_zscaler/actions/lookup_url/action.py [hunk 2]
```diff
@@ -13,7 +14,15 @@ def __init__(self):
                 output=LookupUrlOutput())
 
     def run(self, params={}):
-        lookup_urls = params.get(Input.URLS)
+        urls = params.get(Input.URLS)
+        lookup_urls = []
+        for url in urls:
+            if url and not url.startswith("http"):
+                url = f"http://{url}"
```
Comment (jschipp-r7) [inline zscaler/icon_zscaler/actions/lookup_url/action.py:21]: We should blacklist both the http and https version per ticket.

Comment (jschipp-r7) [inline zscaler/icon_zscaler/actions/lookup_url/action.py:21]: We should lookup both the https and http version of the URL (2 URLs)

File: zscaler/icon_zscaler/actions/blacklist_url/action.py [hunk 3]
```diff
@@ -23,6 +23,9 @@ def run(self, params={}):
         urls = params.get(Input.URLS)
         normalized_urls = []
         for url in urls:
+            if url and not url.startswith("http"):
+                url = f"http://{url}"
```
Comment (jschipp-r7) [inline zscaler/icon_zscaler/actions/blacklist_url/action.py:27]: We should blacklist both the https and http version of the URL (2 URLs)

**Review by jschipp-r7 — CHANGES_REQUESTED** [review]

File: zscaler/icon_zscaler/actions/lookup_url/action.py (hunk 2 above)
Comment (r7-kszczepanskagorna) [reply zscaler/icon_zscaler/actions/lookup_url/action.py:21]: Lookup and blacklist works only for domain. Not url. So we add here http to work with urlparse which get domain part from url. Scheme is not passing to API.

```
lookup_urls.append(
    urlparse(url).hostname
)
```

File: zscaler/icon_zscaler/actions/blacklist_url/action.py (hunk 3 above)
Comment (r7-kszczepanskagorna) [reply zscaler/icon_zscaler/actions/blacklist_url/action.py:27]: Lookup and blacklist works only for domain. Not url. So we add here http to work with urlparse which get domain part from url. Scheme is not passing to API.

```
lookup_urls.append(
    urlparse(url).hostname
)
```

File: zscaler/icon_zscaler/actions/lookup_url/action.py (hunk 2 above)
Comment (jschipp-r7) [reply zscaler/icon_zscaler/actions/lookup_url/action.py:21]: What if URL is https://www.google.com?

Comment (r7-kszczepanskagorna) [reply zscaler/icon_zscaler/actions/lookup_url/action.py:21]:
``` 
urls: ['https://exampletests123.com', 'http://exampletests1234.com', 'exampletests12345.com']
normalized_urls: ['exampletests123.com', 'exampletests1234.com', 'exampletests12345.com']
Starting new HTTPS connection (1): admin.zscalerbeta.net:443
https://admin.zscalerbeta.net:443 "POST /api/v1/authenticatedSession HTTP/1.1" 200 138
Starting new HTTPS connection (1): admin.zscalerbeta.net:443
https://admin.zscalerbeta.net:443 "POST /api/v1/security/advanced/blacklistUrls?action=ADD_TO_LIST HTTP/1.1" 204 0
{"body": {"log": "Connect: Connecting...\nrapid7/Zscaler:1.1.2. Step name: blacklist_url\nurls: ['https://exampletests123.com', 'http://exampletests1234.com', 'exampletests12345.com']\nnormalized_urls: ['exampletests123.com', 'exampletests1234.com', 'exampletests12345.com']\n", "status": "ok", "meta": {}, "output": {"success": true}}, "version": "v1", "type": "action_event"}
``` 
Schema is unused :) I added logger to check those situations :)

File: zscaler/icon_zscaler/actions/blacklist_url/action.py (hunk 3 above)
Comment (jschipp-r7) [reply zscaler/icon_zscaler/actions/blacklist_url/action.py:27]: Shouldn't this be something like to support a user entering both https and https URL and turning them into domain?

```
if url and not (url.startswith("http") or url.startswith("https")):
  url = f"http://{url}"
```

Comment (r7-kszczepanskagorna) [reply zscaler/icon_zscaler/actions/blacklist_url/action.py:27]: @jschipp-r7 This code support http and https, user can provide both and also user can provide domain.

Comment (jschipp-r7) [reply zscaler/icon_zscaler/actions/blacklist_url/action.py:27]: oh, duh `http` is in `https`. Ha, thanks - I had a moment there.

**Review by jschipp-r7 — APPROVED** [review]

**Pull request merged**